cpp/src/stream_compaction/apply_boolean_mask.cu

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include<cudf/types.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/copy_if.cuh>
+#include <cudf/stream_compaction.hpp>
+#include <cudf/detail/stream_compaction.hpp>
+#include <algorithm>
+
+namespace {
+
+// Returns true if the mask is true and valid (non-null) for index i
+// This is the filter functor for apply_boolean_mask
+struct boolean_mask_filter
```

Comment:
As an optimization, consider adding a template bool for whether or not the boolean mask contains nulls.